README.md
# Axiom [![Build Status](https://secure.travis-ci.org/tsujigiri/axiom.png?branch=master)](http://travis-ci.org/tsujigiri/axiom)

Axiom is a micro-framework for building web applications in Erlang.
It is inspired by [Sinatra](http://sinatrarb.com) and built on top of
[Cowboy](https://github.com/extend/cowboy).

## Getting Started

Axiom is built to make creating web applications fast and easy.
A minimal application would look like this:

```erlang
-module(my_app).
-export([start/0, handle/3]).

start() ->
	axiom:start(?MODULE).

handle(<<"GET">>, [<<"hi">>], _Request) ->
	<<"Hello world!">>.

```

This handles requests for `GET /hi` and returns "Hello world!".

The third argument given to the handler is of type `cowboy_req:req()`. Use the
`cowboy_req` module, if you need anything out of the request.

The return value can be a binary string or iolist. So, this also works:

```erlang
handle(<<"GET">>, [<<"hello">>, Who], _Request) ->
	[<<"Hello ">>, Who, <<"!">>].
```

If you want to specify a response status code and/or headers, use a tuple with
either the status code and body or status code, headers and body, in these
respective orders.

Examples:

```erlang
{418, <<"<h1>I'm a teapot!</h1>">>}
```
or
```erlang
{418, [{<<"X-Foo">>, <<"bar">>}], <<"<h1>I'm a teapot!</h1>">>}
```

As a third option a `cowboy_req:req()` can be returned. In this case, to set the
response headers and body, use the `cowboy_req:set_resp_header/3` and
`cowboy_req:set_resp_body/2` functions. To set the status code, use
`axiom:set_resp_status/2`. These functions return a new `cowboy_req:req()` to be
used further and to be returned from `YourHandler:handle/3`.

The full spec of `YourHandler:handle/3` is expected to look like this:

```erlang
handle(Method, Path, Req) -> Body | Req | {Status, Body} | {Status, Headers, Body}.

  Types:
    Method = binary(),
    Path = [PathSegment]
    PathSegment = binary()
    Req = cowboy_req:req()
    Body = iodata()
    Status = non_neg_integer()
    Headers = [Header]
    Header = {binary(), binary()}
```

## Request parameters

To get the request parameters out of the request, you can use the two
handy functions `axiom:params(Req)` and `axiom:param(Name, Req)`.
The first returns a proplist of all parameters, the second one returns
the named parameter's value. Keys and values are binary strings.

## Configuration

`axiom:start/1` has a bigger brother called `axiom:start/2`, taking a
proplist as the second argument. Possible properties and their defaults
are as follows:

```erlang
[
	{nb_acceptors: 100},		% acceptor pool size
	{host, '_'},				% host IP
	{port, 7654},				% host port
	{public, "public"}			% custom path for static files
]
```

## Static Files

Static files are served via the `cowboy_static` handler. By default, every file
in the ./public directory and all its subdirectories will be made accessible via
URL path the same as file's relative path. E.g. the file `./public/about.html`
can be accessed via `GET /about.html`. **Note**: Currently, if the contents of
the ./public subtree change, Axiom needs to be restarted to reflect the change.

You can specify a custom directory via the `public` option.

When you use this feature, it is advisable to start Erlang with the
`+A n` flag. This will start `n` async threads.
Rule of thumb is to use your machine's number of CPU cores.


## Redirects

You can redirect requests with `redirect/2`:

```erlang
handle(<<"GET">>, [<<"foo">>], Req) ->
  Req1 = axiom:redirect("/bar", Req),
  Req;

handle(<<"GET">>, [<<"bar">>], Request) ->
	<<"<h1>Welcome back!</h1>">>.
```

## Templates

Axiom comes with [Django](https://github.com/django/django) template
support via [erlydtl](https://github.com/evanmiller/erlydtl). To make
use of it in your application, create a directory named `templates` and
in it, create a template, e.g. `my_template.dtl`:

```dtl
<h1>Hello {{who}}</h1>
```

In your handler, specify the template to be rendered:

```erlang
handle(<<"GET">>, [<<"hello">>], _Request) ->
	axiom:dtl(my_template, [{who, "you"}]).
```

For convenience, the second argument, a proplist of parameters, can have
atoms, lists or binaries as keys. That way request parameters can be put
in there, without you having to convert them first.

The templates are compiled into modules when `rebar compile` is
called.

To see what else erlydtl can do for you, take a look at
[its project page](https://code.google.com/p/erlydtl/).


## Sessions

Axiom comes with a basic session handler and ets based session store. To
use it, add this tuple to the configuration proplist:

```erlang
{sessions, []}
```

In your handler you can then use
`axiom_session:set(Key, Value, Request)` and
`axiom_session:get(Key, Request)`.

To set attributes for the cookie, storing the session ID, add some
parameters to the session configuration in a tuple with the key
`cookies`:

```erlang
{sessions, [{cookies, [param()]}]}
```

Possible parameters are:

```erlang
param() = {max_age, integer()} |
		  {local_time, calendar:datetime()} |
		  {domain, binary()} |
		  {path, binary()} |
		  {secure, true | false} |
		  {http_only, true | false}
```

The default session store is the `axiom_session_ets` module. You can use
your own by adding a `store` tuple to the sessions tuple:

```erlang
{sessions, [{store, my_session_store, []}]}
```

For implementation details take a look into the `axiom_session_ets`
module.

## Filters

The functions `before_filter/1` and `after_filter/1` can be implemented
to deal with the `cowboy_req:req()` before and after `YourHandler:handle/3`.
When implemented, these are called no matter which `handle` function matches the
request.

In your handler module:

```erlang
before_filter(Req) ->
	% do stuff
	Req.

after_filter(Req) ->
	% do more stuff
	Req.
```

## Errors

### Not Found

To overwrite Axiom's response to 404 errors, just create a catch-all
handler:

```erlang
handle(_Method, _Path, _Req) ->
	{404, <<"nope.">>}.
```

Note that you have to take care of the status code yourself, as
otherwise the default of 200 is sent back to the client.

### Internal Server Error

To handle these yourself, you can implement a function named `error/1`.
The argument is the `cowboy_req:req()` object, otherwise it works like your
`Handler:handle/3` function.

## Streaming

To send a chunked reply, call `axiom:chunk/2` for each chunk:

```erlang
chunk(Data::iodata(), Req::cowboy_req:req()) -> {ok, Req1::cowboy_req:req()}.
```

The returned `cowboy_req:req()` object has to be given as an argument to
subsequent calls to `chunk` and as the return value of your
`Handler:handle/3` function.

To stream data with a Content-Type other than `text/html`, use
`chunk/3`, which has an additional parameter, to be set to the type you
want:

```erlang
chunk(Data::iodata(), Req::cowboy_req:req(), Type::binary()) -> {ok, Req1::cowboy_req:req()}.
```

### Example

```erlang
handle(<<"GET">>, [<<"stream">>], Req) ->
	{ok, Req1} = axiom:chunk(<<"Hello">>, Req, <<"text/plain">>),
	{ok, Req2} = axiom:chunk(<<" world">>, Req1),
	{ok, Req3} = axiom:chunk(<<"!">>, Req2),
	Req3.
```

## Installation

To use it in your OTP application, add this to your `rebar.config`:

```erlang
{lib_dirs, ["deps"]}.
{deps, [
	{'axiom', "0.1.0", {git, "git://github.com/tsujigiri/axiom.git", {tag, "v0.1.0"}}}
]}.
```

then, as usual:

```
rebar get-deps
rebar compile
```

## License

Please take a look at the `LICENSE` file! (tl;dr: it's the MIT License)



CHANGELOG.md
# CHANGELOG

## 0.1.0

* The first argument to `Handler:handle/3` now is a binary string instead of an
  atom.
* The `response` record has been removed. Instead, a tuple with the status code
  and body, or status code, headers and body can be returned from
  `Handler:handle/3`.


test.rb
require 'sinatra'

get '/' do
  params.inspect
end

post '/' do
  params.inspect
end


test/axiom_test_helper.erl
-module(axiom_test_helper).
-compile(export_all).

build_request() ->
	build_request([]).

build_request(Params) ->
	Socket = undefined,
	Transport = ranch_tcp,
	Peer = undefined,
	Method = proplists:get_value(method, Params, <<"GET">>),
	Path = proplists:get_value(path, Params, <<"/">>),
	Query = proplists:get_value('query', Params, undefined),
	Fragment = undefined,
	Version = {1,1},
	Headers = [],
	Host = proplists:get_value(host, Params, <<"example.com">>),
	Port = 80,
	Buffer = <<>>,
	CanKeepalive = false,
	Compress = false,
	OnResponse = undefined,
	cowboy_req:new(Socket, Transport, Peer, Method, Path, Query, Fragment,
		Version, Headers, Host, Port, Buffer, CanKeepalive,
		Compress, OnResponse).



src/axiom_session.erl
-module(axiom_session).
-behaviour(gen_server).

% api
-export([start_link/0, new/1, set/3, get/2, delete/1]).

% callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
	terminate/2, code_change/3]).


-record(state, {session_store, session_state}).

%% API

%% @private
-spec start_link() -> {ok, pid()}.
start_link() ->
	case application:get_env(axiom, sessions) of
		undefined -> ignore;
		{ok, Config} -> gen_server:start_link({local, ?MODULE}, ?MODULE, Config, [])
	end.


%% @doc Creates a new session.
%% 
%% Called by {@link axiom:handle/3} when sessions are configured.
-spec new(cowboy_req:req()) -> cowboy_req:req().
new(Req) ->
	{ExistingId, Req1} = cowboy_req:cookie(<<"session_id">>, Req),
	SessionId = case ExistingId of
		undefined -> new_id();
		_ -> ExistingId
	end,
	Req2 = cowboy_req:set_resp_cookie(
			<<"session_id">>, SessionId, cookie_attributes(), Req1),
	Req3 = cowboy_req:set_meta(session_id, SessionId, Req2),
	ok = gen_server:call(?MODULE, {new, [SessionId, Req3]}),
	Req3.


%% @doc Sets a value in the session.
-spec set(any(), any(), cowboy_req:req()) -> any().
set(Key, Value, Req) ->
	gen_server:call(?MODULE, {set, [Key, Value, Req]}).

%% @doc Gets a value from the session.
-spec get(any(), cowboy_req:req()) -> any().
get(Key, Req) ->
	gen_server:call(?MODULE, {get, [Key, Req]}).

%% @doc Deletes a session.
-spec delete(cowboy_req:req()) -> any().
delete(Req) ->
	gen_server:call(?MODULE, {delete, [Req]}).


%% CALLBACKS

%% @private
-spec init([tuple()]) -> {ok, #state{}}.
init(Config) ->
	{store, SessionStore, SessionConf} = case proplists:lookup(store, Config) of
		none -> {store, axiom_session_ets, []};
		Else -> Else
	end,
	{ok, SessionState} = SessionStore:init(SessionConf),
	{ok, #state{session_state = SessionState, session_store = SessionStore}}.


%% @private
%% @doc Delegates the call to the configured session store module.
-spec handle_call({atom(), [any()]}, {pid(),_}, #state{}) -> {reply, any(), #state{}}.
handle_call({F, A}, _From, State) ->
	M = State#state.session_store,
	SessionState = State#state.session_state,
	{Ret, NewSessionState} = apply(M, F, [SessionState | A]),
	{reply, Ret, State#state{session_state = NewSessionState}}.


%% @private
handle_cast(_Msg, State) ->
	{noreply, State}.

%% @private
handle_info(_Info, State) ->
	{noreply, State}.

%% @private
terminate(_Reason, _State) ->
	ok.

%% @private
code_change(_OldVsn, State, _Extra) ->
	{ok, State}.


%% INTERNAL

%% @private
-spec new_id() -> binary().
new_id() ->
	Data = term_to_binary([make_ref(), now(), random:uniform()]),
	Sha = binary:decode_unsigned(crypto:sha(Data)),
	list_to_binary(lists:flatten(io_lib:format("~40.16.0b", [Sha]))).


%% @private
cookie_attributes() ->
	{ok, Config} = application:get_env(axiom, sessions),
	Attrs = proplists:get_value(cookies, Config, []),
	lists:keystore(path, 1, Attrs, {path, <<"/">>}).



src/axiom_session_ets.erl
-module(axiom_session_ets).

-export([init/1, new/3, set/4, get/3, delete/2]).

-record(state, {tid, conf = []}).

%% @private
-spec init([tuple()]) -> {ok, #state{}}.
init(Conf) ->
	Tid = ets:new(?MODULE, [set, private]),
	{ok, #state{tid = Tid, conf = Conf}}.


%% @doc Store a new session, unless it's not new.
-spec new(#state{}, any(), cowboy_req:req()) -> {ok, #state{}}.
new(State, SessionId, _Req) ->
	ets:insert_new(State#state.tid, {SessionId, []}),
	{ok, State}.


%% @doc Set a value.
-spec set(#state{}, any(), any(), cowboy_req:req()) -> {ok, #state{}}.
set(State, Key, Value, Req) ->
	{SessionId, _} = cowboy_req:meta(session_id, Req),
	Session = case ets:lookup(State#state.tid, SessionId) of
		[{SessionId, Existing}] -> Existing;
		[] -> []
	end,
	NewSession = lists:keystore(Key, 1, Session, {Key, Value}),
	true = ets:insert(State#state.tid, {SessionId, NewSession}),
	{ok, State}.


%% @doc Get a value.
-spec get(#state{}, any(), cowboy_req:req()) -> {any(), #state{}}.
get(State, Key, Req) ->
	{SessionId, _} = cowboy_req:meta(session_id, Req),
	Value = case ets:lookup(State#state.tid, SessionId) of
		[] -> undefined;
		[{_, Session}] -> proplists:get_value(Key, Session)
	end,
	{Value, State}.

%% @doc Delete a session.
-spec delete(#state{}, cowboy_req:req()) -> {ok, #state{}}.
delete(State, Req) ->
	{SessionId, _} = cowboy_req:meta(session_id, Req),
	true = ets:delete(State#state.tid, SessionId),
	{ok, State}.



src/axiom.erl
-module(axiom).
-behaviour(cowboy_http_handler).

% callbacks
-export([init/3, handle/2, terminate/3]).

% api
-export([start/1, start/2, stop/0, params/1, param/2, dtl/1, dtl/2, redirect/2,
	chunk/2, chunk/3, set_resp_status/2, resp_status/1]).

% test proxies
-ifdef(AXIOM_TEST).
-export([process_response_proxy/2]).
-endif.

-record(state, {handler}).

-include_lib("kernel/include/file.hrl").


%% API

%% @doc Starts axiom with default values. Takes the name of the handler
%% module as the argument.
-spec start(module()) -> {ok, pid()}.
start(Handler) ->
	start(Handler, []).


%% @doc Starts axiom with the handler module name as the first and an
%% options proplist as the second argument. Possible options and their
%% default values are as follows:
%%
%% ```
%%	[
%%		{nb_acceptors: 100},	% acceptor pool size
%%		{host, '_'},			% host IP
%%		{port, 7654},			% host port
%%		{public, "public"}		% custom path for static files
%%	]
%% '''
%%
-spec start(module(), [tuple()]) -> {ok, pid()}.
start(Handler, Options) ->
	application:load(axiom),
	lists:map(fun({K,V}) -> application:set_env(?MODULE, K, V) end, Options),
	{ok, Host} = application:get_env(?MODULE, host),
	{ok, Path} = application:get_env(?MODULE, path),
	Dispatch = cowboy_router:compile(
			[{Host, static_dispatch() ++ [{Path, ?MODULE, [Handler]}]}]),
	ok = case application:get_env(axiom, sessions) of
		{ok, _} -> application:start(axiom);
		_ -> ok
	end,
	{ok, NbAcceptors} = application:get_env(axiom, nb_acceptors),
	{ok, Port} = application:get_env(axiom, port),
	cowboy:start_http(axiom_listener, NbAcceptors, [{port, Port}],
		[{env, [{dispatch, Dispatch}]}]
	).


%% @doc Stops axiom.
-spec stop() -> ok.
stop() ->
	case application:get_env(axiom, sessions) of
		undefined -> application:unload(axiom);
		_ -> application:stop(axiom)
	end.


%% @doc Renders a dtl template, takes the template's name as the
%% argument. Templates have to be put in the `templates' directory in
%% your project's root as `my_template.dtl' and rendered with
%% `axiom:dtl(my_template)'.
-spec dtl(atom()) -> iolist();
         (string()) -> iolist().
dtl(Template) ->
	dtl(Template, []).

%% @doc Like `dtl/1' but takes a proplist as the second argument, of
%% which the values are used in the template:
%%
%% For example, a template like this:
%% ```
%% <h1>Hello {{who}}!</h1>
%% '''
%% rendered with:
%% ```
%% axiom:dtl(my_template, [{who, "Helge"}]).
%% '''
%% will result in:
%% ```
%% <h1>Hello Helge!</h1>
%% '''
%%
%% See <a href="https://code.google.com/p/erlydtl/">erlydtl</a> for more
%% information on dtl templates.
-spec dtl(atom(), [tuple()]) -> iolist();
         (string(), [tuple()]) -> iolist().
dtl(Template, Params) when is_atom(Template) ->
	dtl(atom_to_list(Template), Params);

dtl(Template, Params) when is_list(Template) ->
	{ok, Response} =
		apply(list_to_atom(Template ++ "_dtl"), render, [atomify_keys(Params)]),
	Response.


%% @doc Redirect the client to `UrlOrPath'.
-spec redirect(string(), cowboy_req:req()) -> cowboy_req:req().
redirect(UrlOrPath, Req) ->
	{ok, UrlRegex} = re:compile("^https?://"),
	{Url, Req1} = case re:run(UrlOrPath, UrlRegex) of
		{match, _} -> {UrlOrPath, Req};
		nomatch -> assemble_url(UrlOrPath, Req)
	end,
	{Method, Req2} = cowboy_req:method(Req1),
	{Version, Req3} = cowboy_req:version(Req2),
	Status = case {Version, Method} of
		{{1,1}, <<"GET">>} -> 302;
		{{1,1}, _} -> 303;
		_ -> 302
	end,
	Req4 = cowboy_req:set_meta(resp_status, Status, Req3),
	cowboy_req:set_resp_header(<<"Location">>, Url, Req4).


%% @doc Assembles a full URL from a path and a `cowboy_req:req()'.
-spec assemble_url(iolist(), cowboy_req:req()) -> {iolist(), cowboy_req:req()}.
assemble_url(Path, Req) ->
	{Host, Req1} = cowboy_req:host(Req),
	{Port, Req2} = cowboy_req:port(Req1),
	Url = [
		<<"http">>,
		case cowboy_req:get(transport, Req2) of
			ranch_ssl -> <<"s">>;
			_ -> <<>>
		end,
		<<"://">>,
		Host,
		case Port of
			80 -> <<>>;
			Port -> [<<":">>, integer_to_list(Port)]
		end,
		Path
	],
	{Url, Req2}.


%% @doc Extracts the request params from a `cowboy_req:req()'.
-spec params(cowboy_req:req()) -> {[tuple()], cowboy_req:req()}.
params(Req) ->
	{ok, BodyQS, Req1} = cowboy_req:body_qs(Req),
	{QSVals, Req2} = cowboy_req:qs_vals(Req1),
	{QSVals ++ BodyQS, Req2}.


%% @doc extracts a specific param from a `cowboy_req:req()'.
-spec param(binary(), cowboy_req:req()) -> {binary(), cowboy_req:req()}.
param(Param, Req) ->
	{Params, Req1} = params(Req),
	{proplists:get_value(Param, Params), Req1}.


%% @doc Initiates a chunked reply and sends chunked data.
%% If you want to set a `Content-Type' other than `<<"text/html">>', do
%% so with the third argument. Otherwise use {@link chunk/2}.
-spec chunk(iodata(), cowboy_req:req(), binary()) -> {ok, cowboy_req:req()}.
chunk(Data, Req, ContentType) when is_binary(Data) ->
	Req2 = case cowboy_req:get(resp_state, Req) of
		waiting ->
			{ok, Req1} = cowboy_req:chunked_reply(200,
					[{<<"Content-Type">>, ContentType}], Req),
			Pid = spawn(fun() -> stream_loop(Req1) end),
			cowboy_req:set_meta(stream_loop_pid, Pid, Req1);
		chunks -> Req
	end,
	{Loop, Req3} = cowboy_req:meta(stream_loop_pid, Req2),
	Loop ! Data,
	{ok, Req3}.

%% @equiv chunk(Data, Req, <<"text/html">>)
chunk(Data, Req) ->
	chunk(Data, Req, <<"text/html">>).

%% @doc Sets the response status code.
-spec set_resp_status(non_neg_integer(), cowboy_req:req()) -> cowboy_req:req().
set_resp_status(Status, Req) ->
	cowboy_req:set_meta(resp_status, Status, Req).

%% @doc Gets the response status code.
-spec resp_status(cowboy_req:req()) -> {non_neg_integer(), cowboy_req:req()}.
resp_status(Req) ->
	cowboy_req:meta(resp_status, Req).


%% CALLBACKS

%% @private
%% @doc Called by cowboy with a `cowboy_req:req()' and the `state' record
%% we set in {@link init/3}.
-spec handle(cowboy_req:req(), #state{}) -> {ok, cowboy_req:req(), #state{}}.
handle(Req, State) ->
	Handler = State#state.handler,
	Req1 = case application:get_env(axiom, sessions) of
		undefined -> Req;
		{ok, _} -> axiom_session:new(Req)
	end,
	Req2 = cowboy_req:set_resp_header(
			<<"Content-Type">>, <<"text/html">>, Req1),
	Req3 = try
		call_handler(Handler, Req2)
	catch Error:Reason ->
		handle_error(Error, Reason, erlang:get_stacktrace(), Handler, Req)
	end,
	{ok, Req5} = case cowboy_req:get(resp_state, Req3) of
		waiting ->
			{Status, Req4} = cowboy_req:meta(resp_status, Req3, 200),
			cowboy_req:reply(Status, Req4);
		chunks ->
			% wait for the streaming loop to finish sending everything
			{Loop, Req4} = cowboy_req:meta(stream_loop_pid, Req3),
			MonRef = monitor(process, Loop),
			Loop ! terminate,
			receive {'DOWN', MonRef, process, Loop, _Info} -> ok
			after 60000 -> ok
			end,
			{ok, Req4};
		_ -> {ok, Req3}
	end,
	{ok, Req5, State}.


%% @private
%% @doc Called by {@link //cowboy} during initialization.
-spec init({tcp, http}, cowboy_req:req(), [module()]) ->
	{ok, cowboy_req:req(), #state{}}.
init({tcp, http}, Req, [Handler]) ->
	{ok, Req, #state{handler = Handler}}.


%% @private
%% @doc Called by {@link //cowboy} during termination.
-spec terminate(term(), cowboy_req:req(), #state{}) -> ok.
terminate(_Reason, _Req, _State) ->
	ok.

%% INTERNAL FUNCTIONS

%% @private
%% @doc Calls the request handler's `before_filter/1', `after_filter/2'
%% and `handle/3' functions.
-spec call_handler(module(), cowboy_req:req()) -> cowboy_req:req().
call_handler(Handler, Req) ->
	{Method, Req1} = cowboy_req:method(Req),
	{Path, Req2} = cowboy_req:path(Req1),
	Req3 = case erlang:function_exported(Handler, before_filter, 1) of
		true -> Handler:before_filter(Req2);
		false -> Req2
	end,
	{Status, Req4} = cowboy_req:meta(resp_status, Req3),
	case lists:member(Status, [302, 303]) of
		false ->
			SplitPath = case binary:split(Path, <<"/">>, [global, trim]) of
				[<<>> | Rest] -> Rest;
				Else -> Else
			end,
			Req5 = process_response(Handler:handle(Method, SplitPath, Req4), Req4),
			Req6 = case erlang:function_exported(Handler, after_filter, 2) of
				true -> Handler:after_filter(Req5);
				false -> Req5
			end,
			Req6;
		true -> Req4
	end.


%% @private
%% @doc Handles errors in the request handler's `before_filter/1',
%% `after_filter/2' and `handle/3' functions.
-spec handle_error(atom(), any(), [tuple()], module(), cowboy_req:req()) ->
	cowboy_req:req().
handle_error(error, function_clause, [{Handler, handle, _, _} | _], Handler, Req) ->
	{Path, Req1} = cowboy_req:path(Req),
	Req2 = cowboy_req:set_meta(resp_status, 404, Req1),
	Body = dtl(axiom_error_404, [{path, io_lib:format("~p", [Path])}]),
	cowboy_req:set_resp_body(Body, Req2);

handle_error(Error, Reason, Stacktrace, Handler, Req) ->
	Req2 = case erlang:function_exported(Handler, error, 1) of
		true ->
			process_response(Handler:error(Req), Req);
		false -> 
			Req1 = cowboy_req:set_meta(resp_status, 500, Req),
			Body = dtl(axiom_error_500, [
						{error, Error},
						{reason, io_lib:format("~p", [Reason])},
						{stacktrace, format_stacktrace(Stacktrace)}
						]),
			cowboy_req:set_resp_body(Body, Req1)

	end,
	Req2.


%% @private
%% @doc Processes the output of `Handler:handle/3'
-spec process_response({non_neg_integer(), [tuple()], iolist()}, cowboy_req:req()) -> cowboy_req:req();
                      ({non_neg_integer(), iolist()}, cowboy_req:req()) -> cowboy_req:req();
	                  (iolist(), cowboy_req:req()) -> cowboy:req();
	                  (cowboy_req:req(), cowboy_req:req()) -> cowboy_req:req().


process_response({Status, [], Body}, Req) ->
	process_response({Status, Body}, Req);

process_response({Status, [{Key,Value}|Headers], Body}, Req) ->
	Req1 = cowboy_req:set_resp_header(Key, Value, Req),
	process_response({Status, Headers, Body}, Req1);

process_response({Status, Body}, Req) when is_integer(Status) ->
	Req1 = set_resp_status(Status, Req),
	process_response(Body, Req1);

process_response(Body, Req) when is_binary(Body); is_list(Body) ->
	cowboy_req:set_resp_body(Body, Req);

process_response(Req, _) ->
	Req.


%% @private
%% @doc Converts a proplists keys to atoms.
-spec atomify_keys([]) -> [];
                  ([tuple(), ...]) -> [tuple()].
atomify_keys([]) ->
	[];

atomify_keys([Head|Proplist]) ->
	[Key|Tail] = tuple_to_list(Head),
	Key2 = case Key of
		K when is_binary(K) -> list_to_atom(binary_to_list(K));
		K when is_list(K) -> list_to_atom(K);
		K when is_atom(K) -> K
	end,
	[list_to_tuple([Key2|Tail]) | atomify_keys(Proplist)].


%% @private
%% @doc Recursively looks for static files in the `public' directory and tells
%% {@link //cowboy} about it.
-spec static_dispatch() -> [tuple()].
static_dispatch() ->
	{ok, PubDir} = application:get_env(axiom, public),
	filelib:fold_files(PubDir, ".+", true, 
		fun(File, Acc) ->
				RelPath = File -- PubDir,
				Acc1 = [static_dispatch(RelPath, PubDir) | Acc],
				case RelPath =:= "/index.html" of
					true -> [static_dispatch("/", RelPath, PubDir) | Acc1];
					false -> Acc1
				end
		end,
		[]).

%% @private
%% @equiv static_dispatch(Path, Path, Dir)
-spec static_dispatch(string(), string() | tuple()) -> tuple().
static_dispatch(Path, Dir) ->
	static_dispatch(Path, Path, Dir).

%% @private
%% @doc Constructs {@link //cowboy} dispatch configuration for given 
%% static file
-spec static_dispatch(string(), string(), string() | tuple()) -> tuple().
static_dispatch(Path, DiskPath, Dir) ->
	{
		Path,
		cowboy_static,
		[
			{directory, Dir},
			{mimetypes, {fun mimetypes:path_to_mimes/2, default}},
			{file, DiskPath -- "/"}
		]
	}.


%% @private
%% @doc Formats a stacktrace in a more human readable manner.
format_stacktrace([]) ->
	[];

format_stacktrace([H|Stacktrace]) ->
	{M, F, A, Location} = H,
	Line = case proplists:get_value(line, Location) of
		undefined -> "";
		Else -> io_lib:format("~p: ", [Else])
	end,
	File = case proplists:get_value(file, Location) of
		undefined -> "";
		Else2 -> Else2 ++ ":"
	end,
	[io_lib:format("~s~s in ~p:~p/~p", [File, Line, M, F, A]) |
		format_stacktrace(Stacktrace)].


%% @private
-spec stream_loop(cowboy_req:req()) -> ok.
stream_loop(Req) ->
	receive
		terminate -> ok;
		Data ->
			ok = cowboy_req:chunk(Data, Req),
			stream_loop(Req)
	end.


%% test proxies

-ifdef(AXIOM_TEST).

process_response_proxy(Resp, Req) -> process_response(Resp, Req).

-endif.

